### PR TITLE
Clean up remotes's exps

### LINKED ISSUE #6006: Remove pushed experiment from git remote
Push an experiment to a git remote:

```
$ dvc exp push origin myexpe
```

Then:

```
$ dvc exp list origin
master:
   myexpe
```

Currently, only local experiment can be removed with:

```
$ dvc exp remove myexpe
```

but `dvc exp list origin` still shows `myexpe`.

I would like to also be able to remove a pushed experiment on the remote.

### PR DESCRIPTION
**Pull request by karajan1001**
fix #6006 

1. add a new argument `--git-remote` to `dvc exp remove`
2. add some tests for it

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

**Comment by karajan1001** [comment]
@dberenbaum @pmrowla 
Should I support a more granular remove like in `dvc exp push`. The current version will clean up all of the exps in the remove repo.

File: dvc/scm/git/backend/dulwich.py
Comment (karajan1001) [inline dvc/scm/git/backend/dulwich.py:406]: Delete something is also a kind of overwrite. Maybe 

```
if ref in refs:
    local_sha = value
```

Here and add a `--force` flag to it?

is better?

Comment (pmrowla) [reply dvc/scm/git/backend/dulwich.py:406]: If you are explicitly deleting a remote ref, the assumption is that it already exists, and that you want to delete it. I don't see any reason to need a separate `-f/--force` in this case. (This is the same way that `git push -d` works - you don't need to do `git push -f -d`)

Comment (pmrowla) [reply dvc/scm/git/backend/dulwich.py:406]: Also, could you clarify why the `and value != ZERO_SHA` check is needed here?

Comment (karajan1001) [reply dvc/scm/git/backend/dulwich.py:406]: And here what I suggest is that it is better to error out here in `push_refspec` because the `ZERO_SHA` is overwritten to the remote. But we can add the `force=True` flag in the call 

```
repo.scm.push_refspec(remote, None, ref_name)
```
I agreed that we don't need to use `--force` flag here. all of the operations in `dvc exp remove` are dangerous, do not need to remind users.

Comment (pmrowla) [reply dvc/scm/git/backend/dulwich.py:406]: Pushing the `None` source refspec is an explicit remove operation, it shouldn't require an additional `force=True`.

**Comment by dberenbaum** [comment]
> @dberenbaum @pmrowla
> Should I support a more granular remove like in `dvc exp push`. The current version will clean up all of the exps in the remove repo.

Yes, I think granular removal is needed here. My preferred usage would be to make `--git-remote` act like the existing syntax but operating on the remote git repo instead of the local one:

* `dvc exp remove --git-remote origin abc123` removes experiment `abc123` from `origin`.
* `dvc exp remove --git-remote origin --all` removes all experiments from `origin`.
* `dvc exp remove --git-remote origin --queue` does nothing or throws an error.

File: dvc/scm/git/backend/dulwich.py
Comment (dberenbaum) [reply dvc/scm/git/backend/dulwich.py:406]: > I agreed that we don't need to use `--force` flag here. all of the operations in `dvc exp remove` are dangerous, do not need to remind users.

In general, it might be helpful to list the refs found for removal and ask for confirmation by default. That's true for all of `dvc exp remove`, not just with `--git-remote`. I'll open a separate issue for that.

File: dvc/command/experiments.py
Comment (jorgeorpinel) [inline dvc/command/experiments.py:1255]: I don't think we should use `-r` here since that's used for `--remote` in other commands (possibly here later?)
I think argument/flag consistency is a guideline in https://clig.dev/
How about `-g`? or nothing

File: dvc/command/experiments.py
Comment (jorgeorpinel) [inline dvc/command/experiments.py:1256]: Should be something non-repetitive. AFAIR we try to use pseudo types here? In this case `<text>` or maybe `<name>`

File: dvc/command/experiments.py
Comment (jorgeorpinel) [inline dvc/command/experiments.py:1257]: Why GC? Isn't this `exp remove`?

Comment (jorgeorpinel) [reply dvc/command/experiments.py:1257]: Also what's the default? `origin`, I assume. Thanks

**Comment by jorgeorpinel** [comment]
Thinking about it, I'm a bit confused about the command semantics. `dvc push/pull` use the default DVC remote or accepts a `--remote`, `exp push/pull/remove` however take a remote as argument. Seems inconsistent.

Maybe the argument these last accept (should be optional) can be the git remote, resembling `git push/pull`. And then use the default DVC remote or take one via `-r`.

File: dvc/command/experiments.py
Comment (karajan1001) [reply dvc/command/experiments.py:1255]: Can we use `--remote` here? because `dvc exp remove` act only on refs not caches.

File: dvc/command/experiments.py
Comment (karajan1001) [reply dvc/command/experiments.py:1256]: `<remove name or url>` is this too long?

File: dvc/command/experiments.py
Comment (pmrowla) [reply dvc/command/experiments.py:1257]: There is no default, if `--git-remote` is not specified it removes a local experiment ref and not a remote one

Comment (karajan1001) [reply dvc/command/experiments.py:1257]: I think it is better to use "Name or URL of the Git remote to delete all of the experiment references."" 

> Also what's the default? origin, I assume. Thanks
No default for this

**Comment by pmrowla** [comment]
> Thinking about it, I'm a bit confused about the command semantics. `dvc push/pull` use the default DVC remote or accept a `--remote` otherwise. `dvc exp push/pull/remove` however expect a remote as argument. Seems inconsistent.

The mandatory argument for `exp push/pull` is the git remote. DVC data for the experiments is pushed to the default DVC remote if `-r/--remote` is not provided, same as `dvc push/pull`.

> Maybe the argument these last accept should be the git remote (perhaps made optional if the branch has a default upstream), resembling `git push/pull`. And then use the default DVC remote or take one via `-r`. Cc @dberenbaum

Git branch upstream/tracking remotes are specifically only for git branch refs, they don't apply to any other git ref (including experiments). The reason we require explicit git remote arguments is because the "correct" git behavior for "default remotes" will end up being confusing for users.

File: dvc/command/experiments.py
Comment (dberenbaum) [reply dvc/command/experiments.py:1255]: I would defer to @jorgeorpinel, but I think `--git-remote` would be better for consistency and clarity.

File: dvc/command/experiments.py
Comment (dberenbaum) [reply dvc/command/experiments.py:1256]: We are using `metavar="<git_remote>"` already for other `exp push/pull`. If you think it's unhelpful @jorgeorpinel, can we extract to a separate issue and keep as is for now for consistency and to not block this pr?

File: dvc/repo/experiments/utils.py
Comment (pmrowla) [inline dvc/repo/experiments/utils.py:36]: If `name` is an absolute ref (meaning it starts with `refs/`) we don't have to do anything to find the matching ref, we can just return the absolute ref.

Doing the string search/loop iteration is potentially slow, which was why we had the `if name.startswith("refs/")` shortcut checks in the calling functions. The bug in the shortcut implementation was just that we were returning the absolute refs as strings and not as `ExpRefInfo`s (https://github.com/iterative/dvc/issues/6421)

Basically we just want something like:
```py
if name.startswith("refs/"):
    try:
        yield ExpRefInfo.from_ref(name)
    except InvalidExpRefError:
        pass
    return
```

**Review by pmrowla — CHANGES_REQUESTED** [review]

File: dvc/repo/experiments/pull.py
Comment (pmrowla) [inline dvc/repo/experiments/pull.py:41]: All the instances of this should not be removed (see my other comment)

File: dvc/repo/experiments/utils.py
Comment (karajan1001) [reply dvc/repo/experiments/utils.py:36]: Yes, but I guess for the remote refs, this would not cause any performance issue? Because we didn't initialize the remote repo previously?

Comment (karajan1001) [reply dvc/repo/experiments/utils.py:36]: Maybe we should rename it to `exp_ref_info_by_ref`?

Comment (pmrowla) [reply dvc/repo/experiments/utils.py:36]: If the user is providing an absolute `refs/...` string, we can just assume that it exists, either locally or in the remote, and error out later when we try to do whatever we need to with that ref if it turns out to not exist. We don't need to iterate over anything in this case.

It doesn't matter  `ExpRefInfo.from_ref("refs/...")` doesn't matter if it's a local or remote ref, the string is the same either way.

File: dvc/repo/experiments/remove.py
Comment (pmrowla) [inline dvc/repo/experiments/remove.py:100]: `remote` should be an optional parameter optional here and in `_get_exp_ref`

File: dvc/repo/experiments/pull.py
Comment (pmrowla) [inline dvc/repo/experiments/pull.py]:
```suggestion
        return ExpRefInfo.from_ref(exp_name)
```

File: dvc/repo/experiments/remove.py
Comment (pmrowla) [inline dvc/repo/experiments/remove.py:74]: I don't think this check is actually necessary, we can just return the refinfo directly. Also, we can just check if it starts with `refs/` and not `refs/exps`. 

Essentially, if the user gives us a non-exp `refs/...` string, it won't match anything in the rest of this function. Likewise, if `get_ref()` fails, it also means that nothing would match in the rest of this function. So it's faster for us to just return the result of `ExpRefInfo.from_ref()` immediately.

**Comment by karajan1001** [comment]
Fixed #6421 and modified a test line for it.

File: dvc/repo/experiments/utils.py
Comment (pmrowla) [inline dvc/repo/experiments/utils.py:122]: I don't think there will be any use case where we need the list of multiple possible experiments returned by this function. In the event that we wanted multiple matches, we would just use `exp_refs_by_name` directly. If you want to reorganize it this way, this function should probably return a single experiment, and it should raise an appropriate "ambiguous name" error if there is more than one possible matching experiment.

Also, all of the functions in this `utils` module are just doing Git operations, so they take SCM/Git instances as a parameter, rather than a DVC `Repo` instance. It should be the same for this function as well.

To keep parity with the other git libraries we use, it may also make more sense to name it something like `resolve_exp_ref`.

File: tests/func/experiments/test_remote.py
Comment (pmrowla) [inline tests/func/experiments/test_remote.py:154]: Is there a reason for this change? We have a dedicated end-to-end test for pushing and pulling cache (`test_push_pull_cache`)

File: dvc/repo/experiments/utils.py
Comment (pmrowla) [reply dvc/repo/experiments/utils.py:36]: This change can be removed, as discussed we should be doing the check and using `from_ref` elsewhere here. This function is specifically for searching for experiments by the `name` segment of an experiment ref

File: tests/func/experiments/test_remote.py
Comment (karajan1001) [reply tests/func/experiments/test_remote.py:154]: for the #6421 ,

Comment (pmrowla) [reply tests/func/experiments/test_remote.py:154]: So that bug only occurred when we pull cache, but it's not obvious why setting `pull_cache` here should affect whether or not using a full refname works or not.

What we really want for #6421 is a unit test that verifies that the get_exp_ref utils function always returns an `ExpRefInfo`, whether it is passed a shortened name or a full `refs/...` string.

File: dvc/repo/experiments/utils.py
Comment (karajan1001) [reply dvc/repo/experiments/utils.py:122]: Are there any other files for the common function of exp?

File: tests/func/experiments/test_remote.py
Comment (karajan1001) [reply tests/func/experiments/test_remote.py]: Yes but there are two problems.
1. the original function had been removed.
2. we already have a type examine for the new function .
So a functional test or no test and let the issue submitter test him/herself.

Comment (pmrowla) [reply tests/func/experiments/test_remote.py]: Type checking isn't a replacement for a unit test, the new function (`resolve_exp_ref`) should get a unit test instead of this `pull_cache` change.

File: dvc/repo/experiments/utils.py
Comment (pmrowla) [reply dvc/repo/experiments/utils.py:37]: @karajan1001

File: tests/func/experiments/test_utils.py
Comment (pmrowla) [inline tests/func/experiments/test_utils.py:26]:
```suggestion
    assert str(resolve_exp_ref(scm, exp_name, remote)) == str(ref_info)
```

We should be checking that we get the correct ref entirely, not just the name segment.

Actually it would be even better to just define `__hash___()` properly for ExpRefInfo's so that we can just do `if ref_info1 == ref_info2`, but it doesn't have to be done in this PR

Comment (pmrowla) [reply tests/func/experiments/test_utils.py:26]: Also I think this should really be in it's own unit test (it probably also belongs in `tests/unit/experiments/test_utils.py` since it's not a func test). And it should be tested using SCM only, without a full DVC repo instance. Basically something like

```py
@pytest.mark.parametrize("name_only", [True, False])
def test_resolve_exp_ref(tmp_dir, scm, name_only):
    tmp_dir.scm_gen("foo", "foo", commit="init")
    rev = scm.get_rev()
    ref = "/".join([EXPS_NAMESPACE, "ab", "c123", "foo"])
    scm.gitpython.set_ref(ref, rev)
    ref_info = resolve_exp_ref(scm, "foo" if name_only else ref)
    assert isinstance(ref_info, ExpRefInfo)
    assert str(ref_info) == ref
```

Ideally it should also be parametrized for remotes, to do the setup for testing a remote ref you just need
```
scm.push_refspec(remote, ref, ref)
```
(see https://github.com/iterative/dvc/blob/master/tests/unit/scm/test_git.py#L231 for examples on how to set up the git remote)

File: dvc/command/experiments.py
Comment (jorgeorpinel) [reply dvc/command/experiments.py:1256]: Yeah it's a mix in the codebase. At some point I reviewed a lot of metavars and changed their descriptions to metatype annotations but maybe missed some and new ones may be repetitive/unhelpful here. Maybe @skshetry wants to consider this for https://github.com/iterative/dvc/issues/5392 🙂 Up to you guys

File: dvc/command/experiments.py
Comment (jorgeorpinel) [reply dvc/command/experiments.py]: > There is no default

What does `exp remove -g 1234567` do? Errors out? If so OK 👍 thanks

File: dvc/command/experiments.py
Comment (jorgeorpinel) [inline dvc/command/experiments.py:1258]: > Follow-up of https://github.com/iterative/dvc/pull/6471#pullrequestreview-736872715
```suggestion
        help="Name or URL of the Git remote to remove the experiment "
        "from",
```

**Comment by jorgeorpinel** [comment]
> The mandatory argument for exp push/pull is the git remote.

Oh ok thanks @pmrowla I was confused then. That's consistent with `dvc push/pull`.

File: dvc/command/experiments.py
Comment (jorgeorpinel) [inline dvc/command/experiments.py:1257]:
```suggestion
        help="Name or URL of the Git remote to remove the experiment from",
```

File: tests/func/experiments/conftest.py
Comment (pmrowla) [inline tests/func/experiments/conftest.py:106]: This change shouldn't be necessary.

The `git_upstream` fixture creates an git repo that is "upstream" from tmp_dir. So the local `tmp_dir` git repo contains a remote named `upstream` that points to `erepo_dir`. This allows simulating that `tmp_dir` is a local repo, and `erepo_dir` is a server repo.

The `git_downstream` remote creates a repo that is downstream from `tmp_dir`. Meaning that it simulates `tmp_dir` being the "server" repo, and `repo_dir` being a downstream clone of `tmp_dir`. So the git repo in `tmp_dir` contains no remotes at all, and the git repo in `erepo_dir` contains a single remote named `upstream` which points to `tmp_dir`.

File: tests/unit/repo/experiments/test_utils.py
Comment (pmrowla) [inline tests/unit/repo/experiments/test_utils.py:14]: Why is this fixture needed? As far as I can tell it's the same as the existing `git_upstream` fixture?

If the issue is just fixture scope/availability, you can move the git fixtures from `tests/func/experiments/conftest` into the global `tests/dir_helpers` module.

(Or just add the appropriate import into `tests/conftest.py`)

File: tests/conftest.py
Comment (pmrowla) [inline tests/conftest.py:161]: It's probably better to put these in `tests/dir_helpers.py` with the other `tmp_dir` related fixtures

Comment (karajan1001) [reply tests/conftest.py]: I'm just going to push this change.

**Comment by karajan1001** [comment]
If
1. there is no `__init__.py` in the the pytest package ( directory ).
2. Two test file share a same file name.
There would be a failure like this.

```
import file mismatch:
imported module 'test_utils' has this __file__ attribute:
  /home/runner/work/dvc/dvc/tests/unit/repo/experiments/test_utils.py
which is not the same as the test file we want to collect:
  /home/runner/work/dvc/dvc/tests/unit/stage/test_utils.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
```

**Review by pmrowla — APPROVED** [review]

**Comment by pmrowla** [comment]
Looks good, thanks for being patient with all the review requests 👍

File: dvc/command/experiments.py
Comment (dberenbaum) [inline dvc/command/experiments.py:1253]: Should this be part of the mutually exclusive group:

```suggestion
    remove_group.add_argument(
```

Using `-g` with `--queue` doesn't make sense. `dvc exp remove --all -g` doesn't seem to do anything, and I think we discussed that `--all` is probably not wanted for now either since it may be dangerous/unexpected in a collaborative repo.

**Review by dberenbaum — CHANGES_REQUESTED** [review]
One minor issue with the arg parsing. Otherwise looks great!

File: dvc/command/experiments.py
Comment (karajan1001) [reply dvc/command/experiments.py]: If we will not have any of these kinds of feature requests.

**Review by dberenbaum — APPROVED** [review]

**Pull request merged**
